File: src/main/java/com/pauljoda/modularsystems/storage/items/ItemStorageRemote.java
package com.pauljoda.modularsystems.storage.items;

import cofh.api.energy.IEnergyContainerItem;
import com.pauljoda.modularsystems.core.items.BaseItem;
import com.pauljoda.modularsystems.storage.tiles.TileStorageRemote;
import com.teambr.bookshelf.Bookshelf;
import cpw.mods.fml.relauncher.Side;
import cpw.mods.fml.relauncher.SideOnly;
import net.minecraft.creativetab.CreativeTabs;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.util.Vec3;
import net.minecraft.world.World;

import java.util.ArrayList;
import java.util.List;

/**
 * Modular-Systems
 * Created by Dyonovan on 01/08/15
 */
public class ItemStorageRemote extends BaseItem implements IEnergyContainerItem {

    public final int CAPACITY = 10000;
    public final int MAX_IN_OUT = 100;

    public ItemStorageRemote() {
        super("itemStorageRemote", 1);
        setMaxDamage(16);
        setHasSubtypes(true);
    }

    @Override
    public void onCreated(ItemStack itemStack, World world, EntityPlayer entityPlayer) {
        setEnergy(itemStack, 0);
    }

    @Override
    public ItemStack onItemRightClick(ItemStack itemStack, World world, EntityPlayer player) {
        ArrayList<Integer> coords = getCoords(itemStack.getTagCompound());
        if (coords != null) {
            if (world.getTileEntity(coords.get(0), coords.get(1), coords.get(2)) instanceof TileStorageRemote) {
                TileStorageRemote tile = (TileStorageRemote) world.getTileEntity(coords.get(0), coords.get(1), coords.get(2));
                if (isValid(tile, player)) {
                    extractEnergy(itemStack, MAX_IN_OUT, false);
                    player.openGui(Bookshelf.instance, 0, world, tile.getCore().xCoord, tile.getCore().yCoord, tile.getCore().zCoord);
                }
            }
        }

        return itemStack;
    }

    private ArrayList<Integer> getCoords(NBTTagCompound tag) {
        ArrayList<Integer> list = new ArrayList<>();
        if (tag.hasKey("coreX")) {
            list.add(tag.getInteger("coreX"));
            list.add(tag.getInteger("coreY"));
            list.add(tag.getInteger("coreZ"));
            return list;
        }
        return null;
    }

    private boolean isValid(TileStorageRemote tile, EntityPlayer player) {
        if (tile.getCore() != null) {
            if (tile.getWorldObj().provider.dimensionId == player.dimension) {
                if (tile.getCore().canOpen(player)) {
                    int distance = (int)Math.round(Vec3.createVectorHelper(player.posX, player.posY, player.posZ).
                            distanceTo(Vec3.createVectorHelper(tile.xCoord, tile.yCoord, tile.zCoord)));
                    if (distance <= tile.max_distance) {
                        return true;
                    }
                }
            }
        }
        return false;
    }

    @Override
    public int receiveEnergy(ItemStack container, int maxReceive, boolean simulate) {
        if (container.stackTagCompound == null) {
            container.stackTagCompound = new NBTTagCompound();
        }
        int energy = container.stackTagCompound.getInteger("Energy");
        int energyReceived = Math.min(CAPACITY - energy, Math.min(MAX_IN_OUT, maxReceive));

        if (!simulate) {
            energy += energyReceived;
            container.stackTagCompound.setInteger("Energy", energy);
            updateDamage(container);
        }
        return energyReceived;
    }

    @Override
    public int extractEnergy(ItemStack container, int maxExtract, boolean simulate) {
        if (container.stackTagCompound == null || !container.stackTagCompound.hasKey("Energy")) {
            return 0;
        }
        int energy = container.stackTagCompound.getInteger("Energy");
        int energyExtracted = Math.min(energy, Math.min(MAX_IN_OUT, maxExtract));

        if (!simulate) {
            energy -= energyExtracted;
            container.stackTagCompound.setInteger("Energy", energy);
            updateDamage(container);
        }
        return energyExtracted;
    }

    @Override
    public int getEnergyStored(ItemStack container) {
        if (container.stackTagCompound == null || !container.stackTagCompound.hasKey("Energy")) {
            return 0;
        }
        return container.stackTagCompound.getInteger("Energy");
    }

    @Override
    public int getMaxEnergyStored(ItemStack container) {
        return CAPACITY;
    }

    private void updateDamage(ItemStack stack) {
        float r = (float) getEnergyStored(stack) / getMaxEnergyStored(stack);
        int res = 16 - Math.round(r * 16);
        stack.setItemDamage(res);
    }

    void setEnergy(ItemStack container, int energy) {
        if (container.stackTagCompound == null) {
            container.stackTagCompound = new NBTTagCompound();
        }
        container.stackTagCompound.setInteger("Energy", energy);
        updateDamage(container);
    }

    void setFull(ItemStack container) {
        setEnergy(container, CAPACITY);
    }

    @SuppressWarnings("unchecked")
    @Override
    @SideOnly(Side.CLIENT)
    public void getSubItems(Item item, CreativeTabs par2CreativeTabs, List par3List) {
        ItemStack is = new ItemStack(this);
        setFull(is);
        par3List.add(is);

        is = new ItemStack(this);
        setEnergy(is, 0);
        par3List.add(is);
    }

    @SuppressWarnings("unchecked")
    @Override
    @SideOnly(Side.CLIENT)
    public void addInformation(ItemStack itemStack, EntityPlayer player, List list, boolean par4) {
        if (itemStack.stackTagCompound != null) {
            if (itemStack.stackTagCompound.hasKey("coreX")) {
                NBTTagCompound tag = itemStack.stackTagCompound;
                list.add("Receiver Location X:" + tag.getInteger("coreX") + " Y:" + tag.getInteger("coreY") + " Z:" + tag.getInteger("coreZ"));
            } else
                list.add("Unlinked");
            if (itemStack.stackTagCompound.hasKey("Energy"))
                list.add(itemStack.stackTagCompound.getInteger("Energy") + "/" + CAPACITY + " RF");
            else
                list.add("0/" + CAPACITY + " RF");
            return;
        }
        list.add("Unlinked");
    }
}

File: src/main/scala/com/teambr/modularsystems/storage/items/ItemStorageRemote.scala
package com.teambr.modularsystems.storage.items

import cofh.api.energy.IEnergyContainerItem
import com.teambr.bookshelf.Bookshelf
import com.teambr.bookshelf.client.gui.{ GuiColor, GuiTextFormat }
import com.teambr.modularsystems.core.common.items.BaseItem
import com.teambr.modularsystems.storage.tiles.TileStorageRemote
import net.minecraft.creativetab.CreativeTabs
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.item.{Item, ItemStack}
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.util.{Vec3, BlockPos}
import net.minecraft.world.World
import net.minecraftforge.fml.relauncher.{Side, SideOnly}

/**
 * This file was created for Modular-Systems
 *
 * Modular-Systems is licensed under the
 * Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License:
 * http://creativecommons.org/licenses/by-nc-sa/4.0/
 *
 * @author Dyonovan
 * @since August 10, 2015
 */
class ItemStorageRemote extends BaseItem("itemStorageRemote", 1) with IEnergyContainerItem{

    final val CAPACITY = 10000
    final val MAX_IN_OUT = 100

    setMaxDamage(16)
    setHasSubtypes(true)

    override def onCreated(itemStack: ItemStack, world: World, player: EntityPlayer): Unit = setEnergy(itemStack, 0)

    override def onItemRightClick(itemStack: ItemStack, world: World, player: EntityPlayer): ItemStack = {
        if (!world.isRemote) {
            val coords = getCoords(itemStack.getTagCompound)
            if (coords.isDefined) {
                world.getTileEntity(coords.get) match {
                    case tile: TileStorageRemote =>
                        if (isValid(tile, player)) {
                            extractEnergy(itemStack, MAX_IN_OUT, simulate = false)
                            val core = tile.getCore.get.getPos
                            player.openGui(Bookshelf, 0, world, core.getX, core.getY, core.getZ)
                        }
                    case _ =>
                }
            }
        }
        itemStack
    }

    @SideOnly(Side.CLIENT)
    override def getSubItems(item: Item, tab: CreativeTabs, subItems: java.util.List[ItemStack]): Unit = {
        var is = new ItemStack(this)
        setEnergy(is, CAPACITY)
        subItems.asInstanceOf[java.util.List[ItemStack]].add(is)

        is = new ItemStack(this)
        setEnergy(is, 0)
        subItems.asInstanceOf[java.util.List[ItemStack]].add(is)
    }

    @SideOnly(Side.CLIENT)
    override def addInformation(stack: ItemStack, player: EntityPlayer, list: java.util.List[String], boolean: Boolean): Unit = {
        if (stack.hasTagCompound) {
            if (stack.getTagCompound.getLong("Core") != 0) {
                val position = BlockPos.fromLong(stack.getTagCompound.getLong("Core"))
                list.asInstanceOf[java.util.List[String]].add(GuiTextFormat.BOLD + "" +  GuiColor.GREEN +  "Receiver Location:")
                list.asInstanceOf[java.util.List[String]].add("  " + GuiTextFormat.ITALICS + "" +  GuiColor.YELLOW +  "X: " + GuiColor.WHITE + position.getX)
                list.asInstanceOf[java.util.List[String]].add("  " + GuiTextFormat.ITALICS + "" +  GuiColor.YELLOW +  "Y: " + GuiColor.WHITE + position.getY)
                list.asInstanceOf[java.util.List[String]].add("  " + GuiTextFormat.ITALICS + "" +  GuiColor.YELLOW +  "Z: " + GuiColor.WHITE + position.getZ)

            } else list.asInstanceOf[java.util.List[String]].add(GuiColor.RED + "Unlinked")
            if (stack.getTagCompound.getInteger("Energy") != 0) {
                list.asInstanceOf[java.util.List[String]].add(GuiColor.ORANGE + (stack.getTagCompound.getInteger("Energy") + "/" + CAPACITY + " RF"))
            } else list.asInstanceOf[java.util.List[String]].add(GuiColor.RED + "0/" + CAPACITY + " RF")
        } else list.asInstanceOf[java.util.List[String]].add(GuiColor.RED + "Unlinked")
    }

    private def getCoords(tag: NBTTagCompound): Option[BlockPos] = {
        if (tag.hasKey("Core")) {
            return Some(BlockPos.fromLong(tag.getLong("Core")))
        }
        None
    }

    private def isValid(tile: TileStorageRemote, player: EntityPlayer): Boolean = {
        if (tile.getCore.isDefined) {
            if (tile.getWorld.provider.getDimensionId == player.dimension) {
                if (tile.getCore.get.canOpen(player)) {
                    val distance = new Vec3(player.posX, player.posY, player.posZ).distanceTo(
                        new Vec3(tile.getPos.getX, tile.getPos.getY, tile.getPos.getZ)).toInt
                    if (tile.maxDistance >= distance) {
                        return true
                    }
                }
            }
        }
        false
    }

    private def setEnergy(container: ItemStack, energy: Int): Unit = {
        var tag = new NBTTagCompound
        if (container.getTagCompound != null)
            tag = container.getTagCompound
        tag.setInteger("Energy", energy)

        container.setTagCompound(tag)
        updateDamage(container)
    }

    private def updateDamage(stack: ItemStack): Unit = {
        val r = getEnergyStored(stack).toFloat / getMaxEnergyStored(stack)
        val res = 16 - (r * 16).toInt
        stack.setItemDamage(res)
    }

    override def extractEnergy(container: ItemStack, maxExtract: Int, simulate: Boolean): Int = {
        var energy = container.getTagCompound.getInteger("Energy")
        val energyExtracted = Math.min(energy, Math.min(MAX_IN_OUT, maxExtract))

        if (!simulate) {
            energy -= energyExtracted
            setEnergy(container, energy)
        }
        energyExtracted
    }

    override def getEnergyStored(container: ItemStack): Int = container.getTagCompound.getInteger("Energy")

    override def getMaxEnergyStored(container: ItemStack): Int = CAPACITY

    override def receiveEnergy(container: ItemStack, maxReceive: Int, simulate: Boolean): Int = {
        var energy = container.getTagCompound.getInteger("Energy")
        val energyReceived = Math.min(CAPACITY - energy, Math.min(MAX_IN_OUT, maxReceive))

        if (!simulate) {
            energy += energyReceived
            setEnergy(container, energy)
        }
        energyReceived
    }
}
